components/ai_chat/core/browser/engine/engine_consumer_oai.cc

```diff
@@ -269,9 +282,94 @@ void EngineConsumerOAIRemote::GenerateAssistantResponse(
   base::Value::List messages =
       BuildMessages(model_options_, truncated_page_content, selected_text,
                     is_video, conversation_history);
+
+  // Perform the main response request
   api_->PerformRequest(model_options_, std::move(messages),
-                       std::move(data_received_callback),
-                       std::move(completed_callback));
+                       data_received_callback, std::move(completed_callback));
+
+  // For first turn, perform a separate request to generate title
+  if (conversation_history.size() == 1) {
+    GenerateTitle(truncated_page_content, human_input,
+                  std::move(data_received_callback));
+  }
+}
+
+void EngineConsumerOAIRemote::GenerateTitle(
+    const std::string& page_content,
+    const std::string& first_message,
+    GenerationDataCallback data_callback) {
+  std::string conversation_content;
+  if (!page_content.empty()) {
+    conversation_content = base::StrCat(
+        {"Context: ", page_content, "\nQuestion: ", first_message});
+  } else {
+    conversation_content = first_message;
+  }
+
+  std::string truncated_content = conversation_content.substr(0, 1200);
+  base::Value::List title_messages;
+
+  {
+    base::Value::Dict message;
+    message.Set("role", "user");
+    message.Set("content", base::StrCat({kTitlePrompt, truncated_content,
+                                         "</conversation>"}));
+    title_messages.Append(std::move(message));
+  }
+
+  {
+    base::Value::Dict message;
+    message.Set("role", "assistant");
+    message.Set("content",
+                "Here is the title for the above conversation "
+                "in <title> tags:\n<title>");
+    title_messages.Append(std::move(message));
+  }
+
+  api_->PerformRequest(
+      model_options_, std::move(title_messages), base::NullCallback(),
+      base::BindOnce(&EngineConsumerOAIRemote::OnTitleGenerationData,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(data_callback)));
+}
+
+void EngineConsumerOAIRemote::OnTitleGenerationData(
+    GenerationDataCallback data_callback,
+    GenerationResult result) {
+  bool valid_title = result.has_value() && !result->empty();
+  std::string title = "New Chat";
+
+  if (valid_title) {
+    std::string response = base::StrCat({"<title>", *result});
+    std::cerr << "*** Generation: |" << response << "|" << std::endl;
+    size_t start = response.find("<title>");
+    size_t end = response.find("</title>");
+
+    if (start != std::string::npos && end != std::string::npos) {
+      start += 7;  // Length of "<title>"
+      std::string extracted_title =
+          std::string(response.substr(start, end - start));
```

Comment:
Before retrieving the substring, should we impose a maximum, especially since this can go wrong without any maliciousness. e.g. enforce `end` to be the maximum of `end` or `100` characters?

Reply:
And truncate there if it's too long or invalidate the title generation?

Reply:
hard to decide, maybe truncate for now and we can see how it goes